extra_data/calibration.py

```diff
@@ -0,0 +1,1090 @@
+            {'parameter_id': self.parameter_id(k), 'value': str(v)}
+            for k, v in condition.items()]}
+
+    @lru_cache()
+    def detector(self, detector_name):
+        """Detector metadata."""
+
+        resp_detector = Detector.get_by_identifier(
+            self.client, detector_name)
+
+        if not resp_detector['success']:
+            raise CalCatError(resp_detector)
+
+        return {k: resp_detector['data'][k] for k in self.get_detector_keys}
+
+    @lru_cache()
+    def physical_detector_units(self, detector_id, snapshot_at):
+        """Physical detector unit metadata."""
+
+        resp_pdus = PhysicalDetectorUnit.get_all_by_detector(
+            self.client, detector_id, self.format_time(snapshot_at))
+
+        if not resp_pdus['success']:
+            raise CalCatError(resp_pdus)
+
+        return {int(pdu['karabo_da'][-2:]): {
+                    k: pdu[k] for k in self.get_pdu_keys}
+                for pdu in resp_pdus['data']}
+
+    @lru_cache()
+    def calibration_id(self, calibration_name):
+        """ID for a calibration in CalCat."""
+
+        resp_calibration = Calibration.get_by_name(
+            self.client, calibration_name)
+
+        if not resp_calibration['success']:
+            raise CalCatError(resp_calibration)
+
+        return resp_calibration['data']['id']
+
+    @lru_cache()
+    def parameter_id(self, param_name):
+        """ID for an operating condition parameter in CalCat."""
+
+        resp_parameter = Parameter.get_by_name(self.client, param_name)
+
+        if not resp_parameter['success']:
+            raise CalCatError(resp_parameter)
+
+        return resp_parameter['data']['id']
+
+    def closest_ccv_by_time_by_condition(
+        self, detector_name, calibrations, condition,
+        modules=None, event_at=None, snapshot_at=None, metadata=None
+    ):
+        """Query bulk CCV metadata from CalCat.
+
+        This method uses the /get_closest_version_by_detector API
+        to query matching CCVs for PDUs connected to a detector instance
+        in one go. In particular, it automatically includes the PDU as
+        an operating condition parameter to allow for a single global
+        condition rather than PDU-specific ones.
+
+        Args:
+            detector_name (str): Detector instance name.
+            calibrations (Iterable of str): Calibrations to query
+                metadata for.
+            condition (dict): Mapping of parameter name to value.
+            modules (Collection of int or None): List of module numbers
+                or None for all (default).
+            event_at (datetime, date, str or None): Time at which the
+                CCVs should have been valid or None for now (default).
+            snapshot_at (datetime, date, str or None): Time of database
+                state to look at or None for now (default).
+            metadata (dict or None): Mapping to fill for results or
+                None for a new dictionary (default).
```

Comment:
The 'pass in a container to put the output into' pattern seems a little bit ugly. I think I can see why it's there, to simplify combining the retrieved metadata for dark & illuminated conditions for AGIPD & LPD, but I think that can be done simply enough outside this method:

```python
for pdu_idx, pdu_data in new_metadata.items()
    prev_metadata.setdefault(pdu_idx, {}).update(pdu_data)
```

Or invert the nesting so the outer layer is constant types, and then it's just a plain `.update()`. :slightly_smiling_face: